### PR TITLE
docs(Z-flags): Add description for each -Z flag in cargo help

### LINKED ISSUE #5155: Command-line documentation of -Z feature flags
```
$ cargo help
...
    -Z FLAG ...         Unstable (nightly-only) flags to Cargo
...
```
There appears to be no further doc on these here. Cargo's command-line help should at a minimum list which unstable features are available, if not actually print documentation on them.

### PR DESCRIPTION
**Pull request by klausi**
Fixes #5155 

There might be a more elaborate way of adding descriptions to available options and generating help text automatically. For now I think some static text is a good start.

### CODE REVIEW

**Comment by matklad** [comment]
We absolutely must include docs for unstable flags. However, I am not sure that we should add these docs to the message which is printed for `cargo --help`. I think it's important to keep `cargo --help` as small as possible, and target it primarily for the new users. 

What about providing something like `cargo --help -Z`, which would print a verbose description of each feature. Than, `cargo --help` could print something like

```
-Z FLAG ...         Unstable (nightly-only) flags to Cargo, see 'cargo --help -Z' for details
```

**Comment by klausi** [comment]
Good idea! Unfortunately it seems a bit hard to work around Clap's help system to dynamically change help text based on other options. I just introduced another option `--list-Z-flags` for now that returns te help text for the `-Z` options standalone.

**Comment by klausi** [comment]
The cargo help screen for options then looks like this:

```
$ /home/klausi/workspace/cargo/target/debug/cargo help
Rust's package manager

USAGE:
    cargo [OPTIONS] [SUBCOMMAND]

OPTIONS:
    -V, --version           Print version info and exit
        --list              List installed commands
        --explain <CODE>    Run `rustc --explain CODE`
    -v, --verbose           Use verbose output (-vv very verbose/build.rs output)
    -q, --quiet             No output printed to stdout
        --color <WHEN>      Coloring: auto, always, never
        --frozen            Require Cargo.lock and cache are up to date
        --locked            Require Cargo.lock is up to date
    -Z <FLAG>...            Unstable (nightly-only) flags to Cargo, see 'cargo list-Z-flags' for details
        --list-Z-flags      List unstable (nightly-only) flags
    -h, --help              Prints help information
```

And the output of `--list-Z-flags`:

```
$ /home/klausi/workspace/cargo/target/debug/cargo --list-Z-flags
Available unstable (nightly-only) flags.

    avoid-dev-deps      Avoid installing dev-dependencies if possible
    minimal-versions    Install minimal dependency versions instead of maximum
    no-index-update     Do not update the registry, avoids a network request for benchmarking
    offline             Offline mode that does not perform network requests
    print-im-a-teapot   Example option for demonstration purposes
    unstable-options    Allow the usage of unstable options such as --registry

Run with 'cargo -Z [FLAG] [SUBCOMMAND]'
```

**Comment by matklad** [comment]
Hm, apparently, rustc does this as 

    rustc -Z help

Let’s use this as well?

**Comment by klausi** [comment]
Ah yes, that makes sense from a user perspective.

Implementing that seems hard because I don't see a way to display dynamic help text based on other options (-Z) with Clap. Here is my current attempt, which is incomplete and ugly: https://github.com/klausi/cargo/commit/adb0c90f1b8e9527a2828e54694e49a5913d0463

**Comment by klausi** [comment]
One approach would be to call cli().get_matches_from_safe() with removed "help" "--help" and "-h" OS arguments when the display help error case is matched.

Not sure if argument parsing is meant to be called twice in clap or if it has side effects.

**Comment by matklad** [comment]
Oh, sorry, I think I was not clear enough. I meant to have 

   cargo -Z help


as *the only* way to get help about Z flag. It seems to be not that hard to implement? ‘help’ in this case is just a value of the Z option.

So I think that adding something like

    if args.value_of(“unstable-features”) == Some(“help”)

Somewhere around here https://github.com/rust-lang/cargo/blob/715fc78caa832223a3404925a829e18c7832703a/src/bin/cli.rs#L37 should do the trick?

**Comment by matklad** [comment]
Excellent! Let's write a simple test with something like `.with_stdout_countains("-Z unstable-options[..]")`.

File: src/bin/cli.rs [hunk 1]
```diff
@@ -10,6 +10,24 @@ use command_prelude::*;
...
+    if args.value_of("unstable-features") == Some("help") {
+        println!(
+            "
+Available unstable (nightly-only) flags:
+
+    -Z avoid-dev-deps    -- Avoid installing dev-dependencies if possible
+    -Z minimal-versions  -- Install minimal dependency versions instead of maximum
+    -Z no-index-update   -- Do not update the registry, avoids a network request for benchmarking
+    -Z offline           -- Offline mode that does not perform network requests
+    -Z print-im-a-teapot -- Example option for demonstration purposes
```
Comment (matklad) [inline src/bin/cli.rs]: Teapot is just for testing the Cargo itself, I think we should not list it here :-)

**Comment by klausi** [comment]
Ah, figured it out: we need to assert the full text of one line. So the name `with_stdout_contains()` is a bit confusing and should actually be `with_stdout_contains_line()`. Maybe a fix for another time and another PR :)

**Comment by matklad** [comment]
@bors r+

Thanks!

**Comment by klausi** [comment]
Yay, thanks a lot @matklad for your patience and guidance!

**Pull request merged**